tests/baselines/reference/isolatedDeclarationErrors.errors.txt

```diff
@@ -15,6 +16,10 @@ isolatedDeclarationErrors.ts(8,1): error TS9023: Assigning properties to functio
 !!! error TS9023: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     
     const errorOnMissingReturn = () => {}
+                                 ~~~~~~~~
+!!! error TS9007: Function must have an explicit return type annotation with --isolatedDeclarations.

```

Comment:
Was this not an error before for some reason? I can't recall. Also failing on the self check for the same reason.

Reply:
I fixed the self check. This was a bug previously, we were not catching a case that should have been an isolated declarations error (the function had a return with an assertion, but the return type could also be `undefined`). I will have a look at this one tomorrow, but it probably should have been an error? I need to think through our reasoning for not inferring return types on function expressions that have no return again.

Reply:
The overarching reason was that having no return did not imply a `void` return, since CFA internally could instead say it's `never` due to asserts or similar.

For empty funcs, see https://github.com/microsoft/TypeScript/issues/58330, and I think consistency between arrow funcs and func declarations is desirable.

But, this makes me wonder if we should be extracting this fix out separately since it seems like a bug fix that would be good to have without this PR.

Reply:
I am not sure we should ever consider functions without a return type as ok for isolated declarations. Right now we use the type from an assertion in a return. But this is only valid if the return is at the top level. I can make this change in a separate PR,  but are we sure at that point it's still worth it?